odoo/models.py

```diff
@@ -4269,6 +4255,70 @@ def _check_company(self, fnames=None):
                 })
             raise UserError("\n".join(lines))
 
+    def check_access(self, operation: str) -> None:
+        """ Verify that the current user is allowed to perform ``operation`` on
+        all the records in ``self``. The method raises an :class:`AccessError`
+        if the operation is forbidden on the model in general, or on any record
+        in ``self``.
+
+        In particular, when ``self`` is empty, the method checks whether the
+        current user has some permission to perform ``operation`` on the model
+        in general::
+
+            # check that user has some minimal permission on the model
+            records.browse().check_access(operation)
+
+        """
+        if not self.env.su and (result := self._check_access(operation)):
+            exception = result[1]()
+            result[0].invalidate_recordset()
+            raise exception
+
+    def has_access(self, operation: str) -> bool:
+        """ Return whether the current user is allowed to perform ``operation``
+        on all the records in ``self``. The method is fully consistent with
+        method :meth:`check_access` but returns a boolean instead.
+        """
+        if not self.env.su and (result := self._check_access(operation)):
+            result[0].invalidate_recordset()
+            return False
+        return True
+
+    def _filter_access(self, operation: str):
+        """ Return the subset of ``self`` for which the current user is allowed
+        to perform ``operation``. The method is fully equivalent to::
+
+            self.filtered(lambda record: record.has_access(operation))
+
+        """
+        if self and not self.env.su and (result := self._check_access(operation)):
+            result[0].invalidate_recordset()
+            return self - result[0]
+        return self
+
+    def _check_access(self, operation: str) -> tuple[Self, Callable] | None:
+        """ Return ``None`` if the current user has permission to perform
+        ``operation`` on the records ``self``. Otherwise, return a pair
+        ``(records, function)`` where ``records`` are the forbidden records, and
+        ``function`` can be used to create some corresponding exception.
+
+        This method provides the base implementation of
+        methods :meth:`check_access`, :meth:`has_access`
+        and :meth:`_filter_access`. The method may be overridden in order to
+        restrict the access to ``self``.
+        """
+        Access = self.env['ir.model.access']
+        if not Access.check(self._name, operation, raise_exception=False):
+            return self, functools.partial(Access._make_access_error, self._name, operation)
+
+        if any(self._ids):
+            Rule = self.env['ir.rule']
+            domain = Rule._compute_domain(self._name, operation)
+            if domain and (forbidden := self - self.sudo().filtered_domain(domain)):
```

Comment:
Just small perf improvement. Compare len of recordsets and compute the difference only when needed.

Reply:
The performance of the difference is actually negligible compared to the cost of `filtered_domain()`.  And note that having an empty domain already shortcircuits this code.